INT-3816: Superclass mapping for EMExcTypeRouter

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-3816

Add `cause` `supperclass` mapping ability to the `ErrorMessageExceptionTypeRouter`

artembilan committed:
INT-3816: Superclass mapping for EMExcTypeRouter

JIRA: https://jira.spring.io/browse/INT-3816

Add `cause` `supperclass` mapping ability to the `ErrorMessageExceptionTypeRouter`

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/router/ErrorMessageExceptionTypeRouter.java:
```diff
@@ -38,15 +46,28 @@
+							classesCache.put(className, exceptionClass);
+						}
+						catch (ClassNotFoundException e) {
+							throw new IllegalStateException(e);
```
I think we should detect bad keys during initialization rather than runtime, store the cache as a field, and simply log rather than throwing an exception. The runtime exception can be controlled by `resolutionRequired`.

We certainly shouldn't build the cache on every call, but it needs to be cleared/rebuilt whenever the mappings change.

artembilan committed:
Address PR comments:

Introduce `classNameMappings` to store exception classes on the mapping population.
Override all `@ManagedOperation` s to modify `classNameMappings` as well.

artembilan commented:
Pushed.

artembilan commented:
Note: I don't strangle `ClassNotFoundException` because it is really a config (end user) responsibility to provide valid classes for channel mapping.
The `resolutionRequired` is responsible only for the `channelName` resolution. See `AbstractMappingMessageRouter#resolveChannelForName()`.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/router/ErrorMessageExceptionTypeRouter.java:
```diff
@@ -18,20 +18,91 @@
+	}
+
+	private void populateClassNameMapping(Set<String> classNames) {
+		Map<String, Class<?>> newMappings = new ConcurrentHashMap<String, Class<?>>();
```
`newMappings` is not used. You can avoid all this synchronization by atomically replacing `channelMappings` with `newMappings`.

artembilan replied:
Yes, I can. Like it is done in the superclass.
But even that looks for me dangerous. 
See the `synchronized (this.classNameMappings) ` over iteration in the `getChannelKeys()`.
Without that we may end up with the `ConcurrentModificationException`.
Isn't it?

garyrussell replied:
true.

garyrussell commented:
My point is that, today, if someone configures a bad class name it won't cause any problems at runtime.

If we tighten the logic as you suggest, to enforce that the mapping keys actually represent real class names then we at least need a "what's new" entry, thus stating.

artembilan commented:
>  if someone configures a bad class name it won't cause any problems at runtime.

Yes, it doesn't. But from other side that is just an extra useless noise in the `getChannelKeys()`:
```
String causeName = cause.getClass().getName();
if (this.getChannelMappings().keySet().contains(causeName)) {
	mostSpecificCause = causeName;
}
```
So, we check the `Exception` class (already loaded) against some strings. And if the String isn't the real class name it never passes. For me it is silly overhead to configure something non-valid.

I agreed for the "what's new" change and maybe even "Migration Guide" note. But I don't agree to just log the wrong class name. If we don't have the class in the classpath, no reason to allow to configure it for the runtime.

garyrussell commented:
I am fine with an exception as long as we document the behavior change.

artembilan committed:
Address PR comments:

* Remove redundant local variable
* Fix concurrency access around `channelMapping`
* Add `ClassNotFoundException` test for the `ErrorMessageExceptionTypeRouter`
* Add `What's New` note about the `ClassNotFoundException` during init

artembilan commented:
Pushed.
I'll add the `Migration Guide note` after merge.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/router/AbstractMappingMessageRouter.java:
```diff
@@ -173,11 +172,13 @@ public void replaceChannelMappings(Properties channelMappings) {
-		Map<String, String> oldChannelMappings = this.channelMappings;
-		this.channelMappings = newChannelMappings;
+		Map<String, String> oldChannelMappings = new HashMap<String, String>(this.channelMappings);
+		synchronized (this.channelMappings) {
```
While `getChannelMappings()` is now only accessed by `RouterMetrics`, I think it needs to get a copy rather than a hard reference - otherwise the map can change while the caller has a reference.

artembilan replied:
:+1: and even during iteration we end up with the same `ConcurrentModificationException` :smile: 

Any other concerns to push the valuable commit ?

garyrussell replied:
We have to trust that all (and future) subclasses synchronize on the map but I don't think we want the overhead of copying it for every message.

This is why I prefer the atomic update approach.

I think it would be much better to change `setChannelMapings` and `removeChannelMapping` to use an atomic update as well, instead of updating in-place.

artembilan replied:
M-m-m. OK.
So, looks like I didn't understand the approach for the first time.
Will redo.

artembilan committed:
Get rid of `synchronized`, use atomic changes

artembilan commented:
Pushed

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/router/ErrorMessageExceptionTypeRouter.java:
```diff
@@ -18,20 +18,89 @@
+	@ManagedOperation
+	public void setChannelMapping(String key, String channelName) {
+		super.setChannelMapping(key, channelName);
+		this.classNameMappings.put(key, resolveClassFromName(key));
```
Need to create, copy, replace with a new Map here and in remove... (and in the super class) to avoid CME.

artembilan replied:
Hm... I like your approach to business!
But will be "un"-`synchronized` better, than GC attempts around our `volatile`?

artembilan committed:
Additional polishing to avoid `ConcurrentModificationException `

artembilan commented:
Pushed

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/router/AbstractMappingMessageRouter.java:
```diff
@@ -70,9 +68,8 @@
-		newChannelMappings.putAll(channelMappings);
-		this.doSetChannelMappings(newChannelMappings);
+		Map<String, String> newChannelMappings = new ConcurrentHashMap<String, String>(channelMappings);
+		doSetChannelMappings(newChannelMappings);
```
`doSetChannelMappings` still does an in-place clear/replace on the field.

It's not clear why this uses it whereas the others replace the field directly. Perhaps for the logging?

If so, perhaps we should log all changes - mappings won't change frequently.

artembilan replied:
??? Sorry, I am missing anything, or just don't understand you...
```
private void doSetChannelMappings(Map<String, String> newChannelMappings) {
	Map<String, String> oldChannelMappings = this.channelMappings;
	this.channelMappings = newChannelMappings;
	if (logger.isDebugEnabled()) {
		logger.debug("Channel mappings: " + oldChannelMappings + " replaced with: " + newChannelMappings);
	}
}
```
We replace the old field value with the new one. There is no any `clear`.
We use this `doSetChannelMappings()` in those cases when we fully replace one mapping set with another, like this `setChannelMappings()`. 
We can logging for others, but that does not make so much sense.
From other side we even can get rid of this logging as well: you always can use `getChannelMappings()` for diagnostics before setting a new one.

garyrussell replied:
Doh - sorry - my branch didn't have your latest changes :frowning:

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/router/ErrorMessageExceptionTypeRouter.java:
```diff
@@ -39,14 +112,17 @@
 			}
 		}
-		return Collections.singletonList((Object) mostSpecificCause);
+		return mostSpecificCause != null ? Collections.<Object>singletonList(mostSpecificCause) : null;
```
This is not really needed; the caller handles null entries. I'm ok to leave it though.

artembilan replied:
Ah! Yeah! Good catch!
Feel free to fix on merge. 
That was just copy/paste from the `PayloadTypeRouter` :smile:

garyrussell commented on src/reference/asciidoc/router.adoc:
```diff
@@ -959,7 +959,12 @@ As such, please read chapter _<<xml-xpath-routing>>_
-The only difference is that while `PayloadTypeRouter` navigates the instance hierarchy of a payload instance (e.g., `payload.getClass().getSuperclass()`) to find the most specific type/channel mappings, the `ErrorMessageExceptionTypeRouter` navigates the hierarchy of 'exception causes' (e.g., `payload.getCause()`) to find the most specific `Throwable` type/channel mappings.
+The only difference is that while `PayloadTypeRouter` navigates the instance hierarchy of a payload instance (e.g., `payload.getClass().getSuperclass()`) to find the most specific type/channel mappings,
+the `ErrorMessageExceptionTypeRouter` navigates the hierarchy of 'exception causes' (e.g., `payload.getCause()`)
+to find the most specific `Throwable` type/channel mappings and uses `mappingClass.isInstance(cause)` to match the `cause` to its supper class.
```
`s/pp/p` - will fix on merge

artembilan replied:
:smile: I often print this word with two `p` as well as "se**rr**ialization.
Mental block :wink: 

Thank for catching and fix on merge!

garyrussell commented:
Merged as 2d334954372818adbab18a2ae35a6e00c4e4ba8d with polish as discussed.
